python/taichi/lang/matrix.py

```diff
@@ -506,9 +506,11 @@ def diag(dim, val):
         ret = Matrix(dim, dim)
         for i in range(dim):
             for j in range(dim):
-                ret.set_entry(i, j, 0)
-        for i in range(dim):
-            ret.set_entry(i, i, val)
+                if i == j:
+                    ret.set_entry(i, j, val)
+                else:
+                    ret.set_entry(i, j, 0 * val)
+                    # TODO: need a more systematic way to create a "0" with the right type
```

Comment:
I thought it's honored current type system which doesn't provide a `x.dtype` when `x` is a local variable...
FYI I often use `0 * pos[i]` as a meta programming trick to create dim-independent code in TaiGLSL.

Reply:
Sorry about the confusion. The issue is: in the old code, we set all entries to `0`, which is an `i32`. Later when we set it to `1.4`, we store an `f32` to `i32` and this leads to a truncation. 

In the new version, we do not store `0` to the off-diagonal entries. Also note that `0 * f32` leads to a `f32` instead of `i32` :-)